### ISSUE TITLE
Assigning to an attribute with a descriptor

### ISSUE THREAD

**Issue by lambda-abstraction** [microsoft/pyright#9332]
**Describe the bug**
Assigning to an attribute with a descriptor changes the type of getting the value from the descriptor in that scope to what type the value assigned was even if that cant be returned by the descriptors `__get__`

**Code or Screenshots**

```python
from datetime import datetime
from typing import Self, overload, reveal_type


class DateDescriptor:
    @overload
    def __get__(self, instance: None, cls) -> Self:
        ...
    @overload
    def __get__(self, instance: "Thing", cls) -> datetime | None:
        ...
    def __get__(self, instance: "Thing | None", cls = None) -> Self | datetime | None:
        if instance is None:
            return self
        date = instance._date
        if isinstance(date, (float, int)):
            return datetime.fromtimestamp(date)
        else:
            return date
    def __set__(self, instance: "Thing", value: int | float | datetime | None) -> None:
        if isinstance(value, (float, int)):
            instance._date = datetime.fromtimestamp(value)
        elif isinstance(value, datetime):
            instance._date = value
        elif not value:
            ...
        else:
            raise TypeError

class Thing:
    date = DateDescriptor()
    def __init__(self) -> None:
        self._date: int | float | datetime | None = None

thing = Thing()

reveal_type(thing.date) # datetime | None (as expected)

thing.date = 0
reveal_type(thing.date) # Literal[0] (how? the `__get__` can never return that. on mypy its still (datetime | None) as expected)
```

**VS Code extension or command-line**
[pyright playground 1.1.386](https://pyright-play.net/?code=GYJw9gtgBAJghgFwKYIJYSVdAHMIGyIrpIBQokUCAntqgHYDmWEu%2BUAykgDbAA0UMADckIbmDgwBIJCLjcA%2BjWxlSpAMbc4AZ21QAIkX1Jt6kKmwI8ALlJR7UAALDR4yXYcwkwKAoWMUPwAKbR5%2BLHptBDh6dSRrKAA5MHokAU1tAEooAFoAPk4w2wcSqAA6Co97ZxExCRgq2G9ff0CFELCBBiiYuISAIgAVAAsGRn707izcgvhkNAwoAB8klPjGkoqyxq8fPwCEYNDeLsjo2PioIdGmZdXUiagMqABee6Rs-MLeO7niRZWyVSxVK9lQPm65ziWD0QPWoNBMgQAFcQPQoMdgBtPERXhEehcygo-tiwRDtJDekggn8BEFgG4EKcEJlMiCEQ4kaj0X8FkgyhQIHyeqwaURMqSoDxQuyOVAuWjCMgds0-KFDu1MacCX0riMxo8hPJkZcGAQVgyJOalf9MIC1p8CnDZSVwTDKRcgkbuCa6ZbEMzWS6ER64kS-njeSQBeAhSQRdgvcaPpKeG7UBSzlSkz60ja%2BWzJa6s4Tibi3t6TanuG76GACJX4XL7FtqzKi5y4BnMINaEgAKIgcAgNSaHR6fVMWURt6GZDGUzmSx4IISkq7FoMVAajq8R3vYP2THhogJM13f3WqMA954uFqBA3ZhvSeMVdqGRyRTKamPsZlP5sgAYnzEg7jhKAgh0KUAA8VHUZAYAlUg-yYADyygAAGUhPyQeQlD7IJUMYdDkGAqAABlt1EeQAG1MIAXUg4YwAAdwAfioYZMAAA32NoeKeGIoFSWp5RQbkuMQMpBHRCBaGoLAED0KJUG4bhIOvO13lZIA)

**Comment by erictraut** [comment]
The behavior you're seeing here is due to [this documented behavior](https://microsoft.github.io/pyright/#/mypy-comparison?id=type-narrowing-for-asymmetric-descriptors). In this case, the presence of the overload on the `__get__` defeats the logic that distinguishes between a symmetric and asymmetric descriptor. It defaults to symmetric, which is why you're seeing the current behavior. I'll look at extending the logic to handle the overload case.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.387.
